### PR TITLE
feat: add session token injection to provisioner

### LINKED ISSUE #6354: Pre-authenticate `coder` CLI on the Workspace
Hi! I would like to pre-authenticate the `coder` CLI on a Coder workspace. This would enable a smooth experience to shut down or update your Coder environment without having to switch to a local terminal session.

Maybe this could be enabled with a `coder_cli_token` resource in the Terraform provider?

### PR DESCRIPTION
**Pull request by sreya**
This PR adds the plumbing necessary for referencing the `coder_session_token` data source in the TF provider. 

fixes #6354

### CODE REVIEW

**Review by mtojek — COMMENTED** [review]
Nice piece of code, Jon!

File: coderd/apikey/apikey.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	LifetimeSeconds int64
+	Scope           database.APIKeyScope
+	TokenName       string
+	RemoteAddr      string
+}
+
+// Generate generates an API key, returning the key as a string as well as the
+// database representation. It is the responsibility of the caller to insert it
+// into the database.
+func Generate(params CreateParams) (string, database.InsertAPIKeyParams, error) {
```
[the remark this answers is not in the document]
Comment (sreya) [reply coderd/apikey/apikey.go:34]: not a good one, no

File: coderd/provisionerdserver/provisionerdserver.go [hunk 2]
```diff
@@ -1410,6 +1427,51 @@ func InsertWorkspaceResource(ctx context.Context, db database.Store, jobID uuid.
 	return nil
 }
 
+func workspaceSessionTokenName(workspace database.Workspace) string {
+	return fmt.Sprintf("%s_%s_session_token", workspace.OwnerID, workspace.ID)
```
[the remark this answers is not in the document]
Comment (sreya) [reply coderd/provisionerdserver/provisionerdserver.go:1431]: given that it's UUIDs it should uniquely identify a workspace right?

File: coderd/provisionerdserver/provisionerdserver.go [hunk 3]
```diff
@@ -1410,6 +1427,51 @@ func InsertWorkspaceResource(ctx context.Context, db database.Store, jobID uuid.
...
+}
+
+func (server *Server) regenerateSessionToken(ctx context.Context, user database.User, workspace database.Workspace) (string, error) {
```
[the remark this answers is not in the document]
Comment (sreya) [reply coderd/provisionerdserver/provisionerdserver.go:1434]: I considered this and originally wrote it in a transaction, I don't think it's necessary but I can put it in one.

File: coderd/provisionerdserver/provisionerdserver.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/provisionerdserver/provisionerdserver.go:1431]: Oh that's true, I didn't notice IDs...

**Review by mtojek — APPROVED** [review]
LGTM, all comments addressed.

**Pull request merged**